infra/pySourceCredChart/js/lib/pySourceCredChart.js

```diff
@@ -0,0 +1,320 @@
+		this.data = opts.data;
+		this.scores = opts.scores;
+		this.mapping = opts.mapping;
+		this.options = opts.options;
+		
+		//initialize simulation
+		///define variables from data for simulation only
+		///data updates chart and restarts simulation later
+		this.links = this.data[this.mapping.edges];
+		if(that.mapping.source){
+			this.links.forEach(function(d){
+				d.source = d[that.mapping.source];
+				d.target = d[that.mapping.target];
+			});
+		}
+    
+		this.nodes = this.data[this.mapping.nodes];
+		
+		/// define simulation object
+		this.simulation = _simulation(that, this.nodes, this.links);
+    
+		// update chart
+		// this.addZoom();
+		d3updateNodes(that, this.nodes);
+		d3updateEdges(that, this.links);		
+		
+				
+		// Update and restart the simulation.
+		that.simulation.nodes(this.nodes);
+		that.simulation.force("link").links(this.links);
+		that.simulation.alpha(1).restart();
+		
+		function timeIt(){
+		  setTimeout(function(){
+			console.log("call stop");
+			that.simulation.alpha(0).restart();
+			that.simulation.stop();
+			}, 20000);
+		}
+		timeIt();
+	}
+	
+});
+	
+	/// d3 based calculations; should work for both this library and React+d3 UI	
+	function _radius(that, d){
+		if(that.scores){
+			// Use the square of the score as radius, so area will be proportional to score (if score available).
+			// For the Python implementation, the scores are separate from the node so we use the node index to
+			// find the score.
+			const _maxScore = 100;
+			const score = that.scores[d.index];
+			const r = Math.sqrt((score / _maxScore) * MAX_SIZE_PIXELS) + 3;
+			
+			if (!isFinite(r)) {
+			  return 0;
+			}
+			return r;
+		} else {
+			return 5
+			} 
+		
+		}
+	
+	function _simulation(that, nodes, links){
+		
+		// set link force
+		const linkForce = d3.forceLink(links)
+			//.id((d) => d.address)
+			.distance(120);
+			
+		// set charge 
+		const nodeCharge = d3.forceManyBody().strength(-380);
+		
+		// set node collide
+		const nodeCollide = d3.forceCollide().radius((d) => {
+			return _radius(that,d);
+		});
+			
+		//set simulation
+		 const simulation = d3.forceSimulation(nodes)
+			.force("charge", nodeCharge)
+			.force("link", linkForce)
+			.force("collide", nodeCollide)
+			.force("x", d3.forceX())
+		    .force("y", d3.forceY())
+		    .alphaTarget(0.02)
+		    .alphaMin(0.01)
+		    .on("tick", d3ticked);
+			
+		function d3ticked(){
+			that.node.selectAll('.node').attr("cx", function(d) { return d.x; })
+					.attr("cy", function(d) { return d.y; })
+				
+			//TODO: fix arrow marker by moving back based on the node radius
+			that.edge.selectAll('.edge').attr("x1", function(d) { return d.source.x; })
+				.attr("y1", function(d) { return d.source.y; })
+				.attr("x2", function(d) { return d.target.x; })
+				.attr("y2", function(d) { return d.target.y; });
+			}
+			
+		return simulation
+	}
+	
+	/// d3 based DOM functions; trying to mimic React Components
+	function d3setScaffolding(that){
```

Comment:
nit: i would rename it to `setupD3Scaffolding` (or at least `d3SetupScaffolding`)